### ISSUE TITLE
RFC: Improved logging framework

### ISSUE THREAD

**Issue by ryzhyk** [vmware/differential-datalog#618]
Rationale
-----------

Improve logging framework to support emitting whether the log is part of an insert or a delete and the logical timestamp when it was generated.

Current logging mechanism, implemented as a library, emits the same log regardless of the operation, since most differential operators do not "know" the context in which they get invoked. The solution is to embed logging operations in the [inspect](https://docs.rs/differential-dataflow/0.11.0/differential_dataflow/collection/struct.Collection.html#method.inspect) operator.  This operator behaves as a pass-through filter (and thus can be embedded anywhere in the dataflow) and contains an action that can produce a side effect, such as generating a log message.  It takes three arguments: (1) an input record (in case of DDlog, this will contain all variables visible at this point in the rule), (2) weight (>0 = insert, <0 = delete), (3) logical timestamp (in DDlog, this currently corresponds to the number of updates pushed to the program, but will be changed to correspond to the number of transactions committed so far; inside a recursive fragment of the program, the timestamp is a tuple consisting of the global timestamp and iteration number inside the recursive fragment). By wrapping logging statements in an `inspect`, we will be able to add this useful information to the log.

Caveats
-----------
- Differential can generate multiple insertions and deletions for the same timestamp, which may still cause confusing log messages. The number of updates can be reduced by applying the [consolidate](https://docs.rs/differential-dataflow/0.11.0/differential_dataflow/operators/consolidate/index.html) operator before inspect; however, `consolidate` can have a non-trivial cost; in addition, even though it does not change the semantics of the program, it does change its internal behavior, which may be undesirable during debugging.

Syntax
---------

```
R(x, z) :- 
   R1(x, y),
   R2(y, z),
   Inspect debug("Evaluating R: ts=${timestamp}, weight=${weight}, x=${x}, y=${y}, z=${z}").
```
Here, `Inspect` is a new DDlog keyword that maps into DD's `inspect` operator.  The keyword is followed by an arbitrary expression whose type must be `()`.  The expression can access all program variables visible at the given point in the rule, plus two special variables: `ddlog_weight: std.DDWeight` and `ddlog_timestamp: std.DDTimestamp`, where:

```
typedef DDWeight = s64
typedef DDTimestamp = DDTSGlobal{epoch: u64}
                    | DDTSNested{epoch: u64, iteration: u64}
```

**Note**: Once we have support for closures (soon!), there will be no need for the special variables, instead the body of `Inspect` will be a closure: `|ts: DDTimestamp, weight: DDWeight| debug(..)`.

The `Inspect` operator can contain arbitrarily complex computations, e.g.,
```
Inspect {
    if (x > y) {
        debug("Evaluating R: ts=${timestamp}, weight=${weight}, x=${x}, y=${y}, z=${z}")
    };
    for (it in z) {
        debug("Evaluating R: ts=${timestamp}, it=${z}")
    }
}.
```

Questions
------------

- `Inspect` can in principle occur anywhere in the body of a rule; however for implementation simplicity and to avoid Heisenberg effects we may want to disallow `Inspect` between operators that DDlog compiles into the same DD operator, e.g.,:
  ```
  R(q, z) :- 
     R1(x, y),
     R2(y, z),
     // cannot have Inspect here
     x > z,
     // cannot have Inspect here
     var q  = x + y.
     // here is fine
     Inspect debug("Evaluating R: ts=${timestamp}, weight=${weight}, x=${x}, y=${y}, z=${z}").
  ```
  OTOH, this restriction may appear arbitrary to users and may restrict their ability to inspect intermediate points in the computation.

**Comment by ryzhyk** [comment]
Implementation details
------------------------

- Add support for the `Inspect` operator to Rust runtime (`differential_datalog/program.rs`). This change can be implemented and tested purely in Rust without touching the Haskell compiler:
  - Declare new operator here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/rust/template/differential_datalog/program.rs#L477.
  - Implement it here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/rust/template/differential_datalog/program.rs#L1777
  - Testing: add the `Inspect` operator to some of the examples in `differential_datalog/test.rs`. Instead of writing to stdout, accumulate log messages in a data structure that can be checked at the end of the test. At this point, `(cd rust/template && cargo test --all)` should succeed, and we can move on to fixing Haskell code.
- Add support for `Inspect` to the compiler
  - `lib/std.dl`: declare `DDTimestamp` and `DDWeight` types.
  - `Syntax.hs`: Add new operator `RHSInspect` here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/src/Language/DifferentialDatalog/Syntax.hs#L527.  Add new kind of expression context here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/src/Language/DifferentialDatalog/Syntax.hs#L1046.
  - `Parse.hs`: add `Inspect` keyword; extend parsec grammar to parse `Inspect` operator.
  - `NS.hs`: Define variables visible inside the `Inspect` operator in: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/src/Language/DifferentialDatalog/NS.hs#L139, specifically all variables  visible at this point in the rules are visible as read-only variables inside `Inspect`; in addition, `ddlog_timestamp` and `ddlog_weight` variables are visible.
  - `Type.hs`: Handle the new operator in `ctxExpectType`.
  - `Rule.hs`: Update all affected functions in this module.
  - `Validate.hs`: Possibly disallow `Inspect` in some contexts.
  - `DatalogProgram.hs`: Support `Inspect` here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/src/Language/DifferentialDatalog/DatalogProgram.hs#L102.
  - `Compile.hs`: This is the most involved part.  The `Inspect` operator operates on collections (vs arrangements) and thus must be handled similar to `FlatMap` here: https://github.com/vmware/differential-datalog/blob/ec444e4932fa7092d9cc8b1b226a99688604770a/src/Language/DifferentialDatalog/Compile.hs#L1459  The main caveat is that, unlike all other operators, `Inspect` cannot process preceding filters, if any, so they must be compiled into a separate `FilterMap` operator, e.,g, in the following rule:
    ```
    R(q, z) :- 
       R1(x, y),
       R2(y, z),
       x > z,
       var q  = x + y.
       Inspect debug("Evaluating R: ts=${timestamp}, weight=${weight}, x=${x}, y=${y}, z=${z}").
    ```
    there will be three operators: `Join` (for `R1(x,y), R2(y,x)`), `FilterMap` (for `x>z, var q = x + y`), and `Inspect`.
  - Add a test for the new operator to `test/datalog_tests/simple2.dl`.

**Comment by ryzhyk** [comment]
Closed via #653
